sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcDeserializer.scala

```diff
@@ -73,135 +75,157 @@ class OrcDeserializer(
    * Creates a writer to write ORC values to Catalyst data structure at the given ordinal.
    */
   private def newWriter(
-      dataType: DataType, updater: CatalystDataUpdater): (Int, WritableComparable[_]) => Unit =
+      dataType: DataType, reuseObj: Boolean)
+  : (CatalystDataUpdater, Int, WritableComparable[_]) => Unit =
     dataType match {
-      case NullType => (ordinal, _) =>
+      case NullType => (updater, ordinal, _) =>
```

Comment:
@msamirkhan sorry if I rushed to read but why do we need pull this updater out here? If the field writers are created once, we don't have to bother about reusing the updater for nested types as they are just created once and cheap.

Reply:
Only the top level field writers are currently being created once. For each level below, they are created for each data point. Eg, the case for struct did the following:

`case st: StructType => (ordinal, value) =>`
`...`
`    val fieldConverters = st.map(_.dataType).map { dt => newWriter(dt, fieldUpdater) }.toArray`
`...`

i.e., it returns a function, which when called will create relevant fieldConverters. Instead, the field converters should be created outside the function so that they are created only once.

Now, the updaters are not always reused. Eg, under an array or a map (i.e., whenever reuseObj turns to false).